packages/sessions/src/trackers/arweave.ts

```diff
@@ -0,0 +1,569 @@
+        )
+        break
+    }
+
+    const { id, toVersion, toImageHash, executor } = items[0]
+
+    const [data, toConfig] = await Promise.all([
+      fetchItem(id).then(response => response.text()),
+      this.configOfImageHash({ imageHash: toImageHash })
+    ])
+
+    if (!toConfig) {
+      throw new Error(`unable to find to config ${toImageHash} for migration`)
+    }
+
+    const mainModule = new ethers.utils.Interface(MAIN_MODULE_ABI)
+    const [encoded, nonce, signature] = mainModule.decodeFunctionData('execute', data)
+    const transactions = commons.transaction.fromTxAbiEncode(encoded)
+    const subdigest = commons.transaction.subdigestOfTransactions(wallet, chainId, nonce, transactions)
+
+    return {
+      tx: { entrypoint: executor, transactions, nonce, chainId, intent: { id: subdigest, wallet }, signature },
+      fromVersion,
+      toVersion: Number(toVersion),
+      toConfig
+    }
+  }
+
+  saveMigration(_address: string, _signed: migrator.SignedMigration, _contexts: commons.context.VersionedContext): Promise<void> {
+    throw new Error('arweave backend does not support saving migrations')
+  }
+}
+
+async function findItems(
+  filter: { [name: string]: string | string[] },
+  options?: { namespace?: string; pageSize?: number; maxResults?: number }
+): Promise<{ [id: string]: { [tag: string]: string } }> {
+  const namespace = options?.namespace
+  const pageSize = options?.pageSize ?? 100
+  const maxResults = options?.maxResults
+
+  const tags = Object.entries(filter).map(
+    ([name, values]) =>
+      `{ name: "${namespace ? `${namespace}-${name}` : name}", values: [${typeof values === 'string' ? `"${values}"` : values.map(value => `"${value}"`).join(', ')}] }`
+  )
+
+  const edges: Array<{ cursor: string; node: { id: string; tags: Array<{ name: string; value: string }> } }> = []
+
+  for (let hasNextPage = true; hasNextPage && (maxResults === undefined || edges.length < maxResults); ) {
+    const query = `
+      query {
+        transactions(sort: HEIGHT_DESC, ${edges.length ? `first: ${pageSize}, after: "${edges[edges.length - 1].cursor}"` : `first: ${pageSize}`}, tags: [${tags.join(', ')}]) {
+          pageInfo {
+            hasNextPage
+          }
+          edges {
+            cursor
+            node {
+              id
+              tags {
+                name
+                value
+              }
+            }
+          }
+        }
+      }
+    `
+
+    let response: Response
+    while (true) {
+      response = await fetch('https://arweave.net/graphql', {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ query }),
+        redirect: 'follow'
+      })
+      if (response.status !== 429) {
+        break
+      }
+      await new Promise(resolve => setTimeout(resolve, RATE_LIMIT_RETRY_DELAY))
```

Comment:
could you add a console.warn() or something if we hit a 429? 

"arweave.net client received rate limit request, waiting for ${RATE_LIMIT_RETRY_DELAY/1000)} seconds and trying again..."